Add F() Macros

jimmys01 opened the pull request.

jimmys01 committed:
Add F() Macros

crankyoldgit commented on src/IRutils.cpp:
```diff
@@ -365,36 +365,36 @@ std::string resultToSourceCode(const decode_results *results) {
   // Only display the value if the decode type doesn't have an A/C state.
   if (!hasACState(results->decode_type))
     output += " " + uint64ToString(results->value, 16);
```
You missed this one. F(" ");

jimmys01 replied:
Single character strings perhaps is better to be done like chars `' '`?

crankyoldgit replied:
Try it and see. I suspect it won't work as a 'x' is a char and a "x" is a string.
It all depends on the addition(+) method associated with Arduino's `String` type, and c++'s `std::string` type. If string + char works in both cases, I'm more than happy to convert them to chars instead of strings.

crankyoldgit commented on src/IRutils.cpp:
```diff
@@ -365,36 +365,36 @@ std::string resultToSourceCode(const decode_results *results) {
 #if DECODE_AC
       uint16_t nbytes = results->bits / 8;
-      output += "uint8_t state[" + uint64ToString(nbytes) + "] = {";
+      output += String(F("uint8_t state[")) + uint64ToString(nbytes) + String(F("] = {"));
```
Why are you using `String()` here? this will cause the unit tests to break (`String()` is only available in the Arduino Framework.) Does it not compile under the Arduino IDE with out it?

E.g. https://travis-ci.org/markszabo/IRremoteESP8266/jobs/518287553#L7282

jimmys01 replied:
`F("something")+ whatever ` does not compile. With `String(F("something")` it does. If you wish i will revert or refactor those lines
```
output += String(F("uint8_t state["));
output += uint64ToString(nbytes);
output += String(F("] = {"));
```

crankyoldgit replied:
And this is one of the problems with the use of the `F()` macro. :-/
Yeah, if it doesn't compile due to incompatible type issues we'll need to break it down to:

```
output += F("uint8_t state[");
output += uint64ToString(nbytes);
output += F("] = {");
```

How does that work/compile?

jimmys01 replied:
yes that should work. I will do this in the coming days.

crankyoldgit replied:
No worries. Good luck. Take your time.

crankyoldgit commented on src/IRutils.cpp:
```diff
@@ -365,36 +365,36 @@ std::string resultToSourceCode(const decode_results *results) {
       for (uint16_t i = 0; i < nbytes; i++) {
-        output += "0x";
-        if (results->state[i] < 0x10) output += "0";
+        output += String(F("0x"));
```
Ditto.

jimmys01 replied:
There the `String `is not needed :-)

crankyoldgit commented on src/IRutils.cpp:
```diff
@@ -365,36 +365,36 @@ std::string resultToSourceCode(const decode_results *results) {
-                  uint64ToString(results->address, 16) + ";\n";
-        output += "uint32_t command = 0x" +
-                  uint64ToString(results->command, 16) + ";\n";
+        output += String(F("uint32_t address = 0x")) +
```
Again, ditto. Please don't use `String()` unless we have to. (I'll stop flagging these)

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -409,7 +409,7 @@ std::string IRDaikinESP::renderTime(uint16_t timemins) {
   hours = timemins / 60;
   ret = uint64ToString(hours) + ":";
```
F(":")

jimmys01 replied:
String(F(":")) 
;-)

crankyoldgit commented on src/ir_Haier.cpp:
```diff
@@ -304,10 +304,10 @@ std::string IRHaierAC::timeToString(const uint16_t nr_mins) {
-  if (nr_mins / 24 < 10) result += "0";  // Zero pad.
+  if (nr_mins / 24 < 10) result += F("0");  // Zero pad.
   result += uint64ToString(nr_mins / 60);
   result += ":";
```
You missed this.

crankyoldgit replied:
i.e. `F(":")`

crankyoldgit commented on src/ir_Panasonic.cpp:
```diff
@@ -614,7 +614,7 @@ std::string IRPanasonicAc::timeToString(const uint16_t mins_since_midnight) {
 #endif  // ARDUINO
   result += uint64ToString(mins_since_midnight / 60) + ":";
```
Missed one.

crankyoldgit commented on src/ir_Whirlpool.cpp:
```diff
@@ -405,11 +405,11 @@ std::string IRWhirlpoolAc::timeToString(const uint16_t minspastmidnight) {
-  if (hours < 10) result += "0";
+  if (hours < 10) result += F("0");
   result += uint64ToString(hours);
   result += ":";
```
Missed one.

crankyoldgit reviewed (changes requested):
Thanks for the changes. That's a lot of effort. :)
You've use `String()` calls when I think you don't need to. They are to be avoided as they are probably unneeded and they require special handling to allow the code to compile outside of the Arduino Framework. e.g. Unit tests and tools etc. use standard gcc/g++ etc on Linux etc.

jimmys01 committed:
Remove String()

crankyoldgit commented on src/ir_Gree.cpp:
```diff
@@ -373,8 +376,8 @@ std::string IRGreeAC::toString() {
     result += F("Manual");
-  result +=
-      String(F(", Swing Vertical Pos: ")) + uint64ToString(getSwingVerticalPosition());
+	result += F(", Swing Vertical Pos: ");
```
Nitpick: I think the indentation is wrong here (or at least github shows it as wrong)

crankyoldgit replied:
Fixed.

jimmys01 committed:
Missed one

crankyoldgit reviewed (changes requested):
Travis found at least one compilation error:
```
../src/ir_Daikin.cpp: In member function ‘std::string IRDaikin2::toString()’:
../src/ir_Daikin.cpp:1263:35: error: ‘String’ was not declared in this scope
   result += String(", Swing (H): ") + uint64ToString(getSwingHorizontal());
```

and a pile-of-lint issues:

```
$ python cpplint.py --extensions=c,cc,cpp,ino --headers=h,hpp {src,test,tools}/*.{h,c,cc,cpp,hpp,ino} examples/*/*.{h,c,cc,cpp,hpp,ino}

src/ir_Coolix.cpp:378:  Extra space after ( in function call  [whitespace/parens] [4]
src/ir_Daikin.cpp:463:  Extra space after ( in function call  [whitespace/parens] [4]
src/ir_Daikin.cpp:1231:  Extra space after ( in function call  [whitespace/parens] [4]
src/ir_Daikin.cpp:1289:  Tab found; better to use spaces  [whitespace/tab] [1]
src/ir_Daikin.cpp:1330:  Extra space before last semicolon. If this should be an empty statement, use {} instead.  [whitespace/semicolon] [5]
src/ir_Daikin.cpp:1338:  Extra space before last semicolon. If this should be an empty statement, use {} instead.  [whitespace/semicolon] [5]
src/ir_Gree.cpp:379:  Tab found; better to use spaces  [whitespace/tab] [1]
src/ir_Hitachi.cpp:271:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
src/ir_Panasonic.cpp:691:  Extra space after ( in function call  [whitespace/parens] [4]
src/ir_Tcl.cpp:313:  Extra space before last semicolon. If this should be an empty statement, use {} instead.  [whitespace/semicolon] [5]
src/IRutils.cpp:377:  Tab found; better to use spaces  [whitespace/tab] [1]
src/IRutils.cpp:393:  Tab found; better to use spaces  [whitespace/tab] [1]
src/IRutils.cpp:394:  Tab found; better to use spaces  [whitespace/tab] [1]
src/IRutils.cpp:396:  Tab found; better to use spaces  [whitespace/tab] [1]
src/IRutils.cpp:397:  Tab found; better to use spaces  [whitespace/tab] [1]
src/IRutils.cpp:401:  Tab found; better to use spaces  [whitespace/tab] [1]
src/IRutils.cpp:402:  Tab found; better to use spaces  [whitespace/tab] [1]
Total errors found: 17
```
which need to be fixed.

jimmys01 committed:
Lint issues attempt 1

jimmys01 commented:
Time is up for today, thank you for your guidance

crankyoldgit commented:
> Time is up for today, thank you for your guidance

You've done well so far. That's a lot of effort in those commits. :-)

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -1287,7 +1287,7 @@ std::string IRDaikin2::toString() {
   else
     result += F("Off");
-	result += F(", Beep: ");
+    result += F(", Beep: ");
```
Need to reduce the indent by one level for these two statements (i.e. remove two spaces the _beep_ lines)

crankyoldgit replied:
Fixed.

crankyoldgit commented on src/ir_Gree.cpp:
```diff
@@ -376,7 +376,7 @@ std::string IRGreeAC::toString() {
   else
     result += F("Manual");
-	result += F(", Swing Vertical Pos: ");
+    result += F(", Swing Vertical Pos: ");
```
Ditto. Wrong indent level.

crankyoldgit commented:
Only a few linter issues to go:
```
src/ir_Coolix.cpp:378:  Extra space after ( in function call  [whitespace/parens] [4]
src/ir_Daikin.cpp:463:  Extra space after ( in function call  [whitespace/parens] [4]
src/ir_Daikin.cpp:1231:  Extra space after ( in function call  [whitespace/parens] [4]
src/ir_Daikin.cpp:1331:  Extra space before last semicolon. If this should be an empty statement, use {} instead.  [whitespace/semicolon] [5]
src/ir_Daikin.cpp:1339:  Extra space before last semicolon. If this should be an empty statement, use {} instead.  [whitespace/semicolon] [5]
src/ir_Panasonic.cpp:691:  Extra space after ( in function call  [whitespace/parens] [4]
```

These two I suspect are incorrect indentation levels. or when you refactored a single line into multiple ones.
```
src/ir_Daikin.cpp:1288:  If/else bodies with multiple statements require braces  [readability/braces] [4]
src/ir_Gree.cpp:377:  If/else bodies with multiple statements require braces  [readability/braces] [4]
```

The unit tests now can run though and have detected a number of errors. I'll re-review and try to find what problems have been introduced.

crankyoldgit commented on src/ir_Haier.cpp:
```diff
@@ -321,114 +321,122 @@ std::string IRHaierAC::toString() {
-  result += ", Swing: " + uint64ToString(getSwing()) + " (";
+  result += F(", Swing: ");
+  result += uint64ToString(getSwing());
+  result += F(", Swing: ");
```
Mistake: Duplicate "Swing" line. Should be: `result += F(" (");` instead I think.
Breaks the unit tests.
e.g.
```
ir_Haier_test.cpp:363: Failure
Expected equality of these values:
  "Command: 1 (On), Mode: 0 (AUTO), Temp: 25C, Fan: 0 (AUTO), " "Swing: 0 (Off), Sleep: Off, Health: Off, " "Current Time: 00:00, On Timer: Off, Off Timer: Off"
    Which is: "Command: 1 (On), Mode: 0 (AUTO), Temp: 25C, Fan: 0 (AUTO), Swing: 0 (Off), Sleep: Off, Health: Off, Current Time: 00:00, On Timer: Off, Off Timer: Off"
  haier.toString()
    Which is: "Command: 1 (On), Mode: 0 (AUTO), Temp: 25C, Fan: 0 (AUTO), Swing: 0, Swing: Off), Sleep: Off, Health: Off, Current Time: 00:00, On Timer: Off, Off Timer: Off"
```

crankyoldgit replied:
Fixed.

crankyoldgit commented on src/ir_Coolix.cpp:
```diff
@@ -313,92 +313,96 @@ String IRCoolixAC::toString() {
   if (getMode() != kCoolixFan)  // Fan mode doesn't have a temperature.
-    result += ", Temp: " + uint64ToString(getTemp()) + "C";
-  result += ", Zone Follow: ";
+  result += F(", Temp: ");
```
I think this should be:
```
  if (getMode() != kCoolixFan) {  // Fan mode doesn't have a temperature.
    result += F(", Temp: ");
    result += uint64ToString(getTemp());
    result += 'C';
  }
```

Note: This breaks a unit test.
e.g.
```
ir_Coolix_test.cpp:485: Failure
Expected equality of these values:
  "Power: On, Mode: 4 (FAN), Fan: 5 (AUTO), Zone Follow: Off, " "Sensor Temp: Ignored"
    Which is: "Power: On, Mode: 4 (FAN), Fan: 5 (AUTO), Zone Follow: Off, Sensor Temp: Ignored"
  ircoolix.toString()
    Which is: "Power: On, Mode: 4 (FAN), Fan: 5 (AUTO)255C, Zone Follow: Off, Sensor Temp: Ignored"
```

crankyoldgit replied:
Fixed.

crankyoldgit commented on src/ir_Mitsubishi.cpp:
```diff
@@ -638,77 +638,78 @@ String IRMitsubishiAC::toString() {
-  result += ", FAN: ";
+  result += F(", Temp: ");
+  result += uint64ToString(getTemp());
+  result += F("C, Fan: ");
```
You've changed the case of FAN to Fan.
ie. Change the line to:
```
  result += F("C, FAN: ");
```

This currently breaks a UNIT TEST.
Yeah, we could adjust the unit test to in:
```
ir_Mitsubishi_test.cpp:976: Failure
Expected equality of these values:
  "Power: On (HEAT), Temp: 22C, FAN: SILENT, VANE: AUTO, " "Time: 17:10, On timer: 00:00, Off timer: 00:00, Timer: -"
    Which is: "Power: On (HEAT), Temp: 22C, FAN: SILENT, VANE: AUTO, Time: 17:10, On timer: 00:00, Off timer: 00:00, Timer: -"
  irMitsu.toString()
    Which is: "Power: On (HEAT), Temp: 22C, Fan: SILENT, VANE: AUTO, Time: 17:10, On timer: 00:00, Off timer: 00:00, Timer: -"
```
Either one works.

crankyoldgit replied:
Fixed.

crankyoldgit reviewed (changes requested):
A few bugs and typos were introduced that need to be fixed before merge.
Other than those and the few remaining lint issues, it's looking great!

crankyoldgit committed:
Fix some remaining linter issues and unit test failings.

```
src/ir_Coolix.cpp:378:  Extra space after ( in function call  [whitespace/parens] [4]
src/ir_Daikin.cpp:463:  Extra space after ( in function call  [whitespace/parens] [4]
src/ir_Daikin.cpp:1231:  Extra space after ( in function call  [whitespace/parens] [4]
src/ir_Daikin.cpp:1331:  Extra space before last semicolon. If this should be an empty statement, use {} instead.  [whitespace/semicolon] [5]
src/ir_Daikin.cpp:1339:  Extra space before last semicolon. If this should be an empty statement, use {} instead.  [whitespace/semicolon] [5]
src/ir_Panasonic.cpp:691:  Extra space after ( in function call  [whitespace/parens] [4]
src/ir_Daikin.cpp:1288:  If/else bodies with multiple statements require braces  [readability/braces] [4]
src/ir_Gree.cpp:377:  If/else bodies with multiple statements require braces  [readability/braces] [4]
```

```
ir_Haier_test.cpp:363: Failure
Expected equality of these values:
  "Command: 1 (On), Mode: 0 (AUTO), Temp: 25C, Fan: 0 (AUTO), " "Swing: 0 (Off), Sleep: Off, Health: Off, " "Current Time: 00:00, On Timer: Off, Off Timer: Off"
    Which is: "Command: 1 (On), Mode: 0 (AUTO), Temp: 25C, Fan: 0 (AUTO), Swing: 0 (Off), Sleep: Off, Health: Off, Current Time: 00:00, On Timer: Off, Off Timer: Off"
  haier.toString()
    Which is: "Command: 1 (On), Mode: 0 (AUTO), Temp: 25C, Fan: 0 (AUTO), Swing: 0, Swing: Off), Sleep: Off, Health: Off, Current Time: 00:00, On Timer: Off, Off Timer: Off"
```

```
ir_Coolix_test.cpp:485: Failure
Expected equality of these values:
  "Power: On, Mode: 4 (FAN), Fan: 5 (AUTO), Zone Follow: Off, " "Sensor Temp: Ignored"
    Which is: "Power: On, Mode: 4 (FAN), Fan: 5 (AUTO), Zone Follow: Off, Sensor Temp: Ignored"
  ircoolix.toString()
    Which is: "Power: On, Mode: 4 (FAN), Fan: 5 (AUTO)255C, Zone Follow: Off, Sensor Temp: Ignored"
```

```
ir_Mitsubishi_test.cpp:976: Failure
Expected equality of these values:
  "Power: On (HEAT), Temp: 22C, FAN: SILENT, VANE: AUTO, " "Time: 17:10, On timer: 00:00, Off timer: 00:00, Timer: -"
    Which is: "Power: On (HEAT), Temp: 22C, FAN: SILENT, VANE: AUTO, Time: 17:10, On timer: 00:00, Off timer: 00:00, Timer: -"
  irMitsu.toString()
    Which is: "Power: On (HEAT), Temp: 22C, Fan: SILENT, VANE: AUTO, Time: 17:10, On timer: 00:00, Off timer: 00:00, Timer: -"
```
